documenting schema obfuscators (#47)
openshift/must-gather-clean: README.md

# Configuration

A very basic default configuration you can supply as the above `-c` flag for OpenShift can be found
under [examples/openshift_default.yaml](examples/openshift_default.yaml). If you want to obfuscate your domain names (e.g. DNS entries), then you have to adjust the list of `domainNames` to include yours.

In case you don't need networking or SDN information in the must-gather, you can run the configuration under [examples/openshift_omit_network.yaml](examples/openshift_omit_network.yaml). 
This will ignore the largest files that also take a long time to obfuscate.

> The configuration is explained along examples below, a fully supported schema defined in [JSON schema](https://json-schema.org/)

Reviewer: This sentence needs to be reworded. Currently sounds grammatically incorrect. May be a (.) instead of (,)?

can be found in [schema.json](pkg/schema/schema.json) along with more examples and documentation. A more browsable

> alternative can be found here

Reviewer: ```suggestion
alternative can be found on [json-schema.app](https://json-schema.app/view/%23?url=https%3A%2F%2Fraw.githubusercontent.com%2Fopenshift%2Fmust-gather-clean%2Fmain%2Fpkg%2Fschema%2Fschema.json).
```

Author: added

on [json-schema.app](https://json-schema.app/view/%23?url=https%3A%2F%2Fraw.githubusercontent.com%2Fopenshift%2Fmust-gather-clean%2Fmain%2Fpkg%2Fschema%2Fschema.json).

## Obfuscation

The configured obfuscation logic will be applied to each non-omitted file and on a line-by-line basis.

> Let's start with a very minimal obfuscation example:

Reviewer: Examples can be structurally represented as subheadings or titles instead of providing a conversational beginning.

Author: what would be a good title here?

Reviewer: Yes. That works. _MAC address obfuscation example_ or _Example for MAC address obfuscation_ . You can pick whichever sounds better for you in the context.

```
config:
  obfuscate:
  - type: MAC
```

This configuration applied on a `must-gather` will detect all MAC addresses recursively in all the files. 
Since obfuscation is about replacing the found information, the above will simply replace the found MAC address with `xx:xx:xx:xx:xx:xx`. We call this a `Static` replacement, which is the default for all types of obfuscators.

There is another type of replacement called `Consistent`, that can be configured like this:

```
config:
  obfuscate:
  - type: MAC
    replacementType: Consistent
```

This will detect all MAC addresses and replace them with a "consistent" identifier that looks like this `xxx-mac-000001-xxx`. 

> Let's say one of your eth interfaces has the mac address `52:54:00:5e:ee:c6` and was logged, then `must-gather-clean`  will guarantee that it will always be assigned the same obfuscated consistent identifier across all files in a must-gather. 

Reviewer: 'For example' instead of 'Let's say'

That primarily helps our support and engineers to ensure we can still understand and reproduces challenges that you were facing without putting your classified information at risk.

> Let's look at another obfuscation type named `IP`. This type can be used to clean IP addresses, we support both IPv4 and IPv6 except for local interfaces (`127.0.0.1`, `0.0.0.0` and `::1`) that will be preserved.

Reviewer: Same comment as above. To represent as a title instead of 'Let's say' E.g Example for obfuscation type named IP followed by the content from next line.

You can configure this along with the MAC obfuscator like this:

```
config:
  obfuscate:
  - type: MAC
    replacementType: Consistent
  - type: IP
    replacementType: Consistent
```

> On a line-by-line basis, this will always execute the MAC obfuscation first and then the IP obfuscator - we'll go through this behaviour in more detail in the `Chaining obfuscators and side effects` section below. 

Reviewer: ```suggestion
On a line-by-line basis, this will always execute the MAC obfuscation first and then the IP obfuscator - we'll go through this behaviour in more detail in the following `Chaining obfuscators and side effects` section. 
```

Author: added

Another configuration flag that we support for each obfuscation type is the `target`. The target is useful when the confidential information can be found not only in the file content, but also in the folder or file names. 

> This can very frequently happen with IP addresses, for example through node names. You can control that independently for each type like this:

Reviewer: _for example, through node names_

> This can very frequently happen with IP addresses, for example through node names. You can control that independently for each type like this:

Reviewer: ```suggestion
This can very frequently happen with IP addresses, for example through node names. You can control that independently for each type as following:
```

```
config:
  obfuscate:
  - type: MAC
    replacementType: Consistent
    target: FileContents
  - type: IP
    replacementType: Consistent
    target: FilePath
```

As you can see, the MAC obfuscator would work on file content whereas the IP obfuscator would work only on FilePaths. There is a mixed target called `All`, that will obfuscate on both paths and contents. 

> The default if no target is specified is `FileContents`, it is thus always recommended using the IP obfuscator with `target: All` to not accidentally leak IP information through folder names.

Reviewer: ```suggestion
The default, if no target is specified, is `FileContents`, it is thus always recommended using the IP obfuscator with `target: All` to not accidentally leak IP information through folder names.
```

Author: I did actually write this in the beginning and my grammar checker complains: 
![image](https://user-images.githubusercontent.com/85170409/132464495-918c834f-d6ea-4191-84d4-426891e470e6.png)

I'll go your route :)

> The third built-in type of obfuscation is `Domain`, let's take a look how this can be configured:

Reviewer: If there are defined types of obfuscation, then you can bullet/number them in the beginning and provide examples with a section title as suggested above.

```
config:
  obfuscate:
  - type: Domain
    domainNames:
    - "rhcloud.com"
    - "dev.rhcloud.com"
```

As you can see, this type must be customized by supplying domain names. 

> Kubernetes resources are defined along with their domain names (eg "apps.openshift.io/v1") and thus would be automatically recognized as such and obfuscated as a false-positive.

Reviewer: replace eg with e.g. across document (though as per doc guidelines in product docs we use 'for example' and refrain e.g.)

Author: removed eg everywhere

We thus kindly ask the user to supply their confidential domain names manually through the configuration.

The above definition will obfuscate `rhcloud.com` as `domain0000001` (consistent) or as `xxxxxxxxxxxxx` (static). 
Note that this does not include subdomains, they would need to be separately obfuscated. 

> A domain name defined as `staging.rhcloud.com` would only be obfuscated as `staging.domain0000001`, thus, it makes sense to also include all common subdomains as done in the example with "dev".

Reviewer: May be reword a bit to avoid 'makes sense' phrase.
Additionally, _done in the example with "dev"_. Is this referred to `dev.rhcloud.com`? If yes, you can mention the entire url to be specific.

Author: reworded

### Custom Obfuscations

Aside from the above three built-in types to obfuscate, we also offer custom obfuscators that allow users to fine-tune the replacement of certain strings (eg custom auth token formats, confidential domain knowledge or keywords).

> This comes through two custom types, one is replacement via type `Keywords` the other via type `Regex`.

Reviewer: Place both Keywords and Regex in bullets for better readability

#### Keywords

> Let's start by looking into `Keywords` first:

Reviewer: Redundant text since there's already a relevant heading placed?

```
config:
  obfuscate:
  - type: Keywords
    replacement:
       hello: bye
       tomorrow: yesterday
```

This configuration will simply replace the strings on the left-hand side with the values on the right-hand side. 
The `target` variable here is supported as well, so you can also target specific files and obfuscate their name, for example:

```
config:
  obfuscate:
  - type: Keywords

>     target: FilePath

Reviewer: ```suggestion
    target: <FilePath>
```
This would help users to understand that it's a placeholder and distinguish from static words like `Keywords`
